### PR TITLE
Fixes #2281: Make MockedConstruction stubs close on demand

### LINKED ISSUE #2281: MockedConstruction not autoclosed
Hi,
i am using jUnit5 5.7.1 and Mockito 3.9.0, still on Java 8, starting the tests from IntelliJ IDEA 2020.3.4

I have two testclasses (marked with  `@ExtendWith({MockitoExtension.class})`), which each have a test-method, which mocks the construction of MOCKED_CLASS via `@Mock` as parameter.

The second test fails though, with 

> org.junit.jupiter.api.extension.ParameterResolutionException: Failed to resolve parameter [org.mockito.MockedConstruction<MOCKED_CLASS> arg1] in method [public void SECONDLY_INVOCED_TEST_METHOD(org.mockito.MockedConstruction<MOCKED_CLASS>) throws java.lang.Exception]: 
For MOCKED_CLASS, static mocking is already registered in the current thread
>
>To create a new mock, the existing static mock registration must be deregistered
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.resolveParameter(ExecutableInvoker.java:239)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.resolveParameters(ExecutableInvoker.java:183)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.resolveParameters(ExecutableInvoker.java:144)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:96)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
> 	at java.util.ArrayList.forEach(ArrayList.java:1257)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
> 	at java.util.ArrayList.forEach(ArrayList.java:1257)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
> 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
> 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
> 	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
> 	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
> 	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
> 	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
> Caused by: org.mockito.exceptions.base.MockitoException: 
> For com.oda.core.domain.production.customerImport.returns.ImportReturnsTransmissionFile, static mocking is already registered in the current thread
> 
> To create a new mock, the existing static mock registration must be deregistered
> 	at org.mockito.junit.jupiter.MockitoExtension.resolveParameter(MockitoExtension.java:198)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.resolveParameter(ExecutableInvoker.java:216)
> 	... 50 more
> 
> 

I can prevent that from happening, if i add a call to [`ScopedMock#closeOnDemand`](https://www.javadoc.io/static/org.mockito/mockito-core/3.9.0/org/mockito/ScopedMock.html#closeOnDemand--) on the injected `MockedConstruction<MOCKED_CLASS>` at the end of the first invoced test method.
So my conclusion is that the mocked construction was not auto-closed at the end of the method's scope, which is what i expected `@Mock` in combination with `@ExtendWith({MockitoExtension.class})` to do.

Otherwise `@Mock` would be unusable for this, and i would have to use the try-with-ressouces syntax, to make sure the `MockedConstruction` actually gets released, even if the test fails.

### PR DESCRIPTION
**Pull request by temp-droid**
MockedConstruction should behave like MockedStatic when we use @Mock on it.

## Checklist

 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [x] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [x] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [x] Avoid other runtime dependencies
 - [x] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [x] The pull request follows coding style
 - [x] Mention `Fixes #<issue number>` in the description _if relevant_
 - [x] At least one commit should mention `Fixes #<issue number>` _if relevant_

Allows this kind of test to work in multiple test files:
```java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.MockedConstruction;
import org.mockito.MockedStatic;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertNotNull;

@ExtendWith(MockitoExtension.class)
public class Test1 {

    @Test
    void testStatic(@Mock MockedStatic<ClassToMock> mockedStatic) {
        assertNotNull(mockedStatic);
    }

    @Test
    void testMockedConstruction(@Mock MockedConstruction<ClassToMock> mockConstruction) {
//        mockConstruction.closeOnDemand();
        assertNotNull(mockConstruction);
    }
}
```
Otherwise the `mockConstruction` object won't be closed and generate error described in #2281.

### CODE REVIEW

**Review by TimvdLippe — CHANGES_REQUESTED** [review]
Please add the regression test that you mentioned in your PR description to our test suite in https://github.com/mockito/mockito/tree/main/subprojects/junit-jupiter/src/test/java/org/mockitousage

**Comment by temp-droid** [comment]
Hi @TimvdLippe , tests have been added. Feel free to mention if something needs to be changed!

Sorry if I messed up with the merge (not sure why it was needed), I hope we can squash everything before merging to hide it?

**Review by raphw — APPROVED** [review]
Obvious oversight. Good to merge.

File: subprojects/junit-jupiter/src/test/java/org/mockitousage/CloseOnDemandTest.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package org.mockitousage;
```
Comment (TimvdLippe) [inline subprojects/junit-jupiter/src/test/java/org/mockitousage/CloseOnDemandTest.java:1]: This file is missing a license header:

```
/*
 * Copyright (c) 2021 Mockito contributors
 * This program is made available under the terms of the MIT License.
 */
```

You can run `./gradlew build` locally to get it passing. (You might also need to run `./gradlew spotlessApply` to make the formatter happy)

**Review by TimvdLippe — CHANGES_REQUESTED** [review]
Change looks good to me. We can clean up the commit message when we squash-merge, so don't worry about that 😄 

Only a small nit with regard to the missing license. Other than that this is ready to go.

**Review by TimvdLippe — APPROVED** [review]
Thanks for the fix!

**Pull request merged**